packages/dom-adapters/src/caret/utils/useSelectionChange.ts

```diff
@@ -0,0 +1,143 @@
+
+
+/**
+ * Utility composable that watches for document "selection change" event and delegates the provided callbacks to subscribers.
+ */
+export const useSelectionChange = createSingleton(() => {
+  /**
+   * User to iterate over all inputs and check if selection is related to them.
+   */
+  const inputsWatched: InputWithCaret[] = [];
+
+  /**
+   * WeakMap that stores subscribers for each input.
+   */
+  const subscribers = new WeakMap<InputWithCaret, Subscriber>();
+
+  /**
+   * Checks if selection is related to input
+   *
+   * @param selection - changed document selection
+   * @param input - input to check
+   */
+  function isSelectionRelatedToInput(selection: Selection | null, input: InputWithCaret): boolean {
+    if (!selection) {
+      return false;
+    }
+
+    const range = selection.getRangeAt(0);
+
+    return input.contains(range.startContainer) ?? false;
+  }
+
+  /**
+   * Handler for document "selection change" event.
+   */
+  function onDocumentSelectionChanged(): void {
+    const selection = document.getSelection();
+
+    /**
+     * Iterate over all subscribers WeakMap and call their callbacks.
+     */
+    inputsWatched.forEach((input) => {
+      const subscriber = subscribers.get(input);
+
+      if (subscriber && isSelectionRelatedToInput(selection, input)) {
+        subscriber.callback.call(subscriber.context, selection);
+      }
+    });
+  }
+
+  /**
+   * Subscribe on "selection change" event.
+   *
+   * @param input - input to watch caret change
+   * @param callback - callback that will be called on "selection change" event
+   * @param context - context of the callback
+   */
+  function on(input: InputWithCaret, callback: Subscriber['callback'], context: Subscriber['context']): void {
+    /**
+     * Add input to the list of watched inputs.
+     */
+    inputsWatched.push(input);
+
+    /**
+     * Store subscription
+     */
+    subscribers.set(input, {
+      callback,
+      context,
+    });
+  }
+
+  /**
+   * Unsubscribe from "selection change" event.
+   *
+   * @param input - input to remove subscription
+   */
+  function off(input: InputWithCaret): void {
+    subscribers.delete(input);
+    inputsWatched.splice(inputsWatched.indexOf(input), 1);
+  }
+
+  /**
+   * Initialize document selection change watcher.
+   */
+  function init(): void {
+    /**
+     * We use single for document "selection change" event and delegate the provided callbacks to subscribers.
+     */
+    document.addEventListener('selectionchange', onDocumentSelectionChanged);
+  }
+
+  /**
+   * Destroy document selection change watcher.
+   */
+  function destroy(): void {
+    document.removeEventListener('selectionchange', onDocumentSelectionChanged);
+
+    inputsWatched.forEach((input) => {
+      const subscriber = subscribers.get(input);
+
+      if (subscriber) {
+        off(input);
+      }
+    });
+  }
+
+  init();
+
+  return {
+    on,
+    off,
+    destroy,
```

Comment:
Once destroyed, there'll be no way to re-init as `createSingletone` will always return the same instance

Reply:
yes

Reply:
I've exported `init` method as well. The destroy flow is not designed yet, so it will be implemented once we can understand it